src/main/java/org/folio/rest/impl/CheckinHelper.java

```diff
@@ -47,19 +49,59 @@
 
     // 1. Get piece records from storage
     return retrievePieceRecords(checkinPieces)
-      // 2. Update items in the Inventory if required
+      // 2. Check locationId presence for pieces related to
+      // POLine with createInventory "Instance, Holding" or "Instance, Holding, Item"
+      .thenCompose(this::checkLocationId)
+      // 3. Update items in the Inventory if required
       .thenCompose(this::updateInventoryItems)
-      // 3. Update piece records with checkIn details which do not have
+      // 4. Update piece records with checkIn details which do not have
       // associated item
       .thenApply(this::updatePieceRecordsWithoutItems)
-      // 4. Update received piece records in the storage
+      // 5. Update received piece records in the storage
       .thenCompose(this::storeUpdatedPieceRecords)
-      // 5. Update PO Line status
+      // 6. Update PO Line status
       .thenCompose(this::updatePoLinesStatus)
-      // 6. Return results to the client
+      // 7. Return results to the client
       .thenApply(piecesGroupedByPoLine -> prepareResponseBody(checkinCollection, piecesGroupedByPoLine));
   }
 
+  /**
+   * Checks locationId presence for pieces related to POLine with
+   * createInventory = "Instance, Holding" or "Instance, Holding, Item".
+   *
+   * @return {@link CompletableFuture} which holds map with PO line id as key
+   *         and list of corresponding pieces as value
+   */
+  private CompletableFuture<Map<String, List<Piece>>> checkLocationId(Map<String, List<Piece>> piecesRecords) {
+    getPoLines(piecesRecords)
+      .thenAccept(poLines -> {
+        for(PoLine poLine : poLines) {
+          List<Piece> pieces = piecesRecords.get(poLine.getId());
+          for (Piece piece : pieces) {
+            // Check if locationId doesn't presented in piece from request and retrieved from storage
+            // Corresponding piece from collection
+            CheckInPiece pieceFromRq = checkinPieces.get(poLine.getId()).get(piece.getId());
+            if (pieceFromRq.getLocationId() == null && piece.getLocationId() == null) {
```

Comment:
The API can be used to check-in piece or revert already checked-in piece. I believe we do not need to force location id to be provided if piece needs to become `Expected` (and related item "On order" if available)